drivers/serial/uart_nrfx_uarte.c

```diff
@@ -550,69 +1094,71 @@ static int uarte_instance_init(struct device *dev,
 		 * otherwise such operation may result in HardFault or RAM
 		 * corruption.
 		 */
-		nrf_uarte_tx_buffer_set(uarte, data->tx_buffer, 0);
+		nrf_uarte_tx_buffer_set(uarte, data->int_driven->tx_buffer, 0);
 		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTTX);
 
 		/* switch off transmitter to save an energy */
 		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STOPTX);
 	}
-#endif /* UARTE_INTERRUPT_DRIVEN */
+
+#if IS_ENABLED(CONFIG_UART_ASYNC_API)
+	if (data->async) {
+		return uarte_nrfx_init(dev);
+	}
+#endif
+	/* Enable receiver and transmitter */
+	nrf_uarte_enable(uarte);
+
+	nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_ENDRX);
+
+	nrf_uarte_rx_buffer_set(uarte, &data->rx_data, 1);
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
 
 	return 0;
 }
 
 #define UART_NRF_UARTE_DEVICE(idx)					       \
-	DEVICE_DECLARE(uart_nrfx_uarte##idx);				       \
-	UARTE_##idx##_CREATE_TX_BUFF;					       \
+	UARTE_INT_DRIVEN(idx);						       \
+	UARTE_ASYNC(idx);						       \
 	static struct uarte_nrfx_data uarte_##idx##_data = {		       \
-		.uart_config = {					       \
-			.baudrate =					       \
-				DT_NORDIC_NRF_UARTE_UART_##idx##_CURRENT_SPEED,\
-			.data_bits = UART_CFG_DATA_BITS_8,		       \
-			.stop_bits = UART_CFG_STOP_BITS_1,		       \
-			.parity = UARTE_##idx##_NRF_PARITY_BIT		       \
-					  == NRF_UARTE_PARITY_INCLUDED	       \
-				  ? UART_CFG_PARITY_EVEN		       \
-				  : UART_CFG_PARITY_NONE,		       \
-			.flow_ctrl = UARTE_##idx##_NRF_HWFC_CONFIG	       \
-					     == NRF_UARTE_HWFC_ENABLED	       \
-				     ? UART_CFG_FLOW_CTRL_RTS_CTS	       \
-				     : UART_CFG_FLOW_CTRL_NONE,		       \
-		},							       \
-		UARTE_##idx##_DATA_INIT					       \
+		UARTE_CONFIG(idx),					       \
+		COND_CODE_1(IS_ENABLED(CONFIG_UART_##idx##_ASYNC),	       \
+			(.async = &uarte##idx##_async,),		       \
+			())						       \
+		COND_CODE_1(IS_ENABLED(CONFIG_UART_##idx##_INTERRUPT_DRIVEN),  \
+			(.int_driven = &uarte##idx##_int_driven,),	       \
+			())						       \
 	};								       \
 	static const struct uarte_nrfx_config uarte_##idx##_config = {	       \
 		.uarte_regs = (NRF_UARTE_Type *)			       \
 			DT_NORDIC_NRF_UARTE_UART_##idx##_BASE_ADDRESS,	       \
-		UARTE_##idx##_CONFIG_RTS_CTS,				       \
-		UARTE_##idx##_CONFIG_INIT				       \
+		.rts_cts_pins_set = IS_ENABLED(UARTE_##idx##_CONFIG_RTS_CTS),  \
 	};								       \
 	static int uarte_##idx##_init(struct device *dev)		       \
 	{								       \
 		const struct uarte_init_config init_config = {		       \
 			.pseltxd = DT_NORDIC_NRF_UARTE_UART_##idx##_TX_PIN,    \
 			.pselrxd = DT_NORDIC_NRF_UARTE_UART_##idx##_RX_PIN,    \
-			UARTE_##idx##_NRF_RTS_CTS_PINS,			       \
+			UARTE_NRF_RTS_CTS_PINS(idx),			       \
 		};							       \
-		UARTE_##idx##_INTERRUPTS_INIT();			       \
-		return uarte_instance_init(dev,				       \
-				&init_config,				       \
-				UARTE_##idx##_INTERRUPT_DRIVEN);	       \
+		COND_CODE_1(UART_ISR_ENABLED,				       \
```

Comment:
This will cause that interrupts will be connected and enabled for all UARTE instances if at least one of them will be using interrupt driven or asynchronous API. I guess it was not the intention.